pkg/utils/mime.go

```diff
@@ -0,0 +1,118 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"mime/multipart"
+	"net/mail"
+	"net/textproto"
+	"strings"
+
+	admapi "github.com/awslabs/amazon-eks-ami/nodeadm/api"
+)
+
+type MIMEContentType string
+
+const (
+	mimeBoundary      = "//"
+	mimeVersionHeader = "MIME-Version: 1.0"
+
+	MIMEContentTypeShellScript = `text/x-shellscript; charset="us-ascii"`
+	MIMEContentTypeNodeConfig  = "application/" + admapi.GroupName
+	MIMEContentTypeMultipart   = `multipart/mixed; boundary="` + mimeBoundary + `"`
+)
+
+type MIMEEntry struct {
+	ContentType MIMEContentType
+	Content     string
+}
+
+type MIMEArchive []MIMEEntry
+
+func NewMIMEArchive(content string) (MIMEArchive, error) {
+	archive := MIMEArchive{}
+	if content == "" {
+		return archive, nil
+	}
+	reader, err := archive.getReader(content)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		p, err := reader.NextPart()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, fmt.Errorf("parsing content, %w", err)
+		}
+		slurp, err := io.ReadAll(p)
+		if err != nil {
+			return nil, fmt.Errorf("parsing content, %s, %w", string(slurp), err)
+		}
+		archive = append(archive, MIMEEntry{
+			ContentType: MIMEContentType(p.Header.Get("Content-Type")),
+			Content:     string(slurp),
+		})
+	}
+	return archive, nil
+}
+
+func (ma MIMEArchive) Serialize() (string, error) {
+	buffer := bytes.Buffer{}
+	writer := multipart.NewWriter(&buffer)
+	if err := writer.SetBoundary(mimeBoundary); err != nil {
+		return "", err
+	}
+	buffer.WriteString(mimeVersionHeader + "\n")
+	buffer.WriteString(fmt.Sprintf("Content-Type: %s\n\n", MIMEContentTypeMultipart))
+	for _, entry := range ma {
+		partWriter, err := writer.CreatePart(textproto.MIMEHeader{
+			"Content-Type": []string{string(entry.ContentType)},
+		})
+		if err != nil {
+			return "", fmt.Errorf("creating multi-part section for entry, %w", err)
+		}
+		_, err = partWriter.Write([]byte(entry.Content))
+		if err != nil {
+			return "", fmt.Errorf("writing entry, %w", err)
+		}
+	}
+	writer.Close()
```

Comment:
I had the same question for this writer in the eksbootstrap code, does this writer close need to be `defer`-ed?

Reply:
Funny you ask, fought this for a little while but if you defer it this actually doesn't work because we grab the output buffer string before closing and never terminate the mime archive. So no, don't want to defer here.

Reply:
> So no, don't want to defer here

We still don't close in the error case, right? Should we close if we hit an error still? My understanding is that if we don't close when we should, we leak.

Reply:
If we don't leak, we should comment. Because -- to the naked eye -- this looks like a bug.

Reply:
I can check but I wouldn't think so, this is just handling some string manipulation, not handling file resources or anything. I can check but I imagine the resource would be garbage collected.

Reply:
You are right though that my instinct is that I see close and think it should close whenever the function exists. While I don't think that's the case here, if I can confirm it I'll add a comment noting that.

Reply:
Took a quick peek at the library code, it just handles writing the termination string and setting a flag in the writer. No reason to close it for the error path.